### PR TITLE
GH-91375: Port `_asyncio` static types to heap types and module state

### PR DESCRIPTION
**Pull request by kumaraditya303**
Co-authored-by: Erlend E. Aasland <erlend.aasland@protonmail.com>

* Issue: gh-91375

### CODE REVIEW

File: Modules/_asynciomodule.c [hunk 1]
```diff
@@ -1703,16 +1795,24 @@ FutureIter_throw(futureiterobject *self, PyObject *const *args, Py_ssize_t nargs
...
+{
+    Py_CLEAR(it->future);
+    return 0;
+}
+
 static PyObject *
 FutureIter_close(futureiterobject *self, PyObject *arg)
 {
-    Py_CLEAR(self->future);
+    FutureIter_clear(self);
```
Comment (erlend-aasland) [inline Modules/_asynciomodule.c]:
```suggestion
    (void)FutureIter_clear(self);
```

File: Modules/_asynciomodule.c [hunk 2]
```diff
@@ -3261,28 +3413,59 @@ new_running_loop_holder(PyObject *loop)
...
+PyRunningLoopHolder_clear(PyRunningLoopHolder *rl)
+{
+    Py_CLEAR(rl->rl_loop);
+    return 0;
+}
+
+
+static int
+PyRunningLoopHolder_traverse(PyRunningLoopHolder *rl, visitproc visit,
+                              void *arg)
```
Comment (erlend-aasland) [inline Modules/_asynciomodule.c]:
```suggestion
                             void *arg)
```

**Review by erlend-aasland — COMMENTED** [review]
Thanks for filling in the missing pieces. Looks good!

We should be able to get rid of `module_initialized` now that we've got multi-phase init.
Also, the `module_free(NULL)` at the end of `module_init()` should now be removed.

(BTW, I added two nit suggestions to clean up some style mess introduced by me.)

File: Modules/_asynciomodule.c (hunk 1 above)
Comment (kumaraditya303) [reply Modules/_asynciomodule.c]: Fixed

File: Modules/_asynciomodule.c (hunk 2 above)
Comment (kumaraditya303) [reply Modules/_asynciomodule.c]: Fixed

**Comment by kumaraditya303** [comment]
> We should be able to get rid of module_initialized now that we've got multi-phase init.
Also, the module_free(NULL) at the end of module_init() should now be removed.

Yes, I thought I had done that but apparently not, might have been messed up in rebasing. Fixed now.

**Review by erlend-aasland — APPROVED** [review]

**Comment by erlend-aasland** [comment]
Kumar: Do you think there's a chance of getting one of the other core devs to do a review? (Andrew or maybe Eric?)

**Comment by erlend-aasland** [comment]
> Andrew is not active ATM because of the war, Eric is busy with the other stuff so I don't think he will have time for this.

Yeah, I feared that.

Do you have a chance to check if there's a performance hit? (I'd be very surprised if there is one.)

**Comment by ericsnowcurrently** [comment]
I'm going to take a look, but it might not be immediately.

**Comment by erlend-aasland** [comment]
> I'm going to take a look, but it might not be immediately.

Great, thanks!

**Review by gvanrossum — COMMENTED** [review]
Wow, this goes on and on. I'm not sure if anyone could review it -- I'm sure I can't. I am tempted to just trust Erlend, Kumar, the compiler, and the unit test suite, and merge it.

However, there's a merge conflict that needs to be resolved first...

**Comment by kumaraditya303** [comment]
> However, there's a merge conflict that needs to be resolved first...

Fixed merge conflict.

**Comment by erlend-aasland** [comment]
Most of the changes are trivial stuff like passing the module state around to various support and helper functions.

**Comment by kumaraditya303** [comment]
I am thinking of merging this by next week so that it gets in the next alpha. @ericsnowcurrently If you are not able to review by then, I can do a follow up PR for your review if that's okay with you.

**Comment by kumaraditya303** [comment]
Merged, thanks @erlend-aasland

**Pull request merged**